keanu-project/src/main/java/io/improbable/keanu/plating/loop/LoopBuilder.java

```diff
@@ -0,0 +1,180 @@
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.DoubleProxyVertex;
+import io.improbable.keanu.vertices.generic.nonprobabilistic.If;
+
+public class LoopBuilder {
+    private final Collection<Vertex> initialState;
+    private ImmutableMap.Builder<VertexLabel, VertexLabel> customMappings = ImmutableMap.builder();
+    private int maxLoopCount = Loop.DEFAULT_MAX_COUNT;
+    private boolean throwWhenMaxCountIsReached = true;
+
+    <V extends Vertex<?>> LoopBuilder(Collection<V> initialState) {
+        this.initialState = ImmutableList.copyOf(initialState);
+    }
+
+    /**
+     * An optional method to override the default value
+     *
+     * @param maxCount - the max number of times the loop can run
+     * @return self
+     */
+    public LoopBuilder atMost(int maxCount) {
+        this.maxLoopCount = maxCount;
+        return this;
+    }
+
+    /**
+     * An optional method to override the default behaviour
+     * If the maximum loop count is exceeded, it will log a warning instead of throwing
+     *
+     * @return self
+     */
+    public LoopBuilder dontThrowWhenMaxCountIsReached() {
+        this.throwWhenMaxCountIsReached = false;
+        return this;
+    }
+
+    /**
+     * An optional method to add custom mappings
+     *
+     * @param inputLabel  - the label assigned to the ProxyVertex in frame t
+     * @param outputLabel - the label assigned to a Vertex in frame t-1 which will become the ProxyVertex's parent
+     * @return self
+     */
+    public LoopBuilder mapping(VertexLabel inputLabel, VertexLabel outputLabel) {
+        customMappings.put(inputLabel, outputLabel);
+        return this;
+    }
+
+    /**
+     * A mandatory method to specify the condition
+     *
+     * @param conditionSupplier - a lambda that creates and returns a new BoolVertex
+     * @return the next stage builder
+     */
+    public LoopBuilder2 whilst(Supplier<BoolVertex> conditionSupplier){
+        return whilst(plate -> conditionSupplier.get());
+    }
+
+    /**
+     * A mandatory method to specify the condition
+     *
+     * @param conditionFunction - a lambda that takes the current Plate and creates and returns a new BoolVertex
+     * @return the next stage builder
+     */
+    public LoopBuilder2 whilst(Function<Plate, BoolVertex> conditionFunction){
+        return new LoopBuilder2(initialState, conditionFunction, customMappings.build(), maxLoopCount, throwWhenMaxCountIsReached);
+    }
+
+    public class LoopBuilder2 {
+        private final Collection<Vertex> initialState;
+        private final Function<Plate, BoolVertex> conditionFunction;
+        private final Map<VertexLabel, VertexLabel> customMappings;
+        private final int maxLoopCount;
+        private final boolean throwWhenMaxCountIsReached;
+        private final VertexLabel VALUE_OUT_WHEN_ALWAYS_TRUE_LABEL = new VertexLabel("loop_value_out_when_always_true");
```

Comment:
We're really relying on the labels here to describe the structure of the graph. Should these special labels that describe structure be some other type of meta data on the vertex?

Reply:
I've changed it now so that Plates keeps its own mapping from VertexLabel to Vertex and doesn't rely on the Vertex's member variable VertexLabel.